Add debug utilities for Vulkan

darksylinc opened the pull request:
Features:
- Debug-only tracking of objects by type. See get_driver_allocs_by_object_type et al.
 - Debug-only Breadcrumb info for debugging GPU crashes and device lost
 - Performance report per frame from get_perf_report
- Some VMA calls had to be modified in order to insert the necessary memory callbacks

Functionality marked as "debug-only" is only available in debug or dev builds.

Misc fixes:
 - Shaders are more aggressively unloaded
 - Early break optimization in RenderingDevice::uniform_set_create

============================

The work was performed by collaboration of TheForge and Google. I am merely splitting it up into smaller PRs and cleaning it up.


# PR comments

This is the first PR from the TheForge changed. This one is the most important one because:

1. It sets the foundation for the rest of the upcoming changes.
2. 99% of these changes should be harmless. They are not compiled in non-debug and non-dev builds.
3. They're quite verbose in terms of lines of code, so getting this out of the way really reduces the mental load.

## Update

Thanks everyone for the help! Most of the issues have been addressed. Almost all CI passes now. Windows now compiles.

**3 outstanding things remain:**

1. ~Turning breadcrumbs into a single `uint32_t`. I want the enum to still be defined in rendering_device_commons.h and document it better.~ Done.
2. ~CI is complaining about [draw_list_begin](https://github.com/godotengine/godot/actions/runs/8862390044/job/24335322837) changing its signature (this is correct; although I'm not sure if the error is triggering because the bindings are wrong?). Help & Tips on this are appreciated.~
3. ~I've been talking with Dario, and it appears TheForge added `RenderingDevice::shader_destroy_modules` which must be called by hand to free (unused) Shader modules after creating the PSO. We both agree it seems like adding a shotgun to the API just to save few kilobytes of RAM (maybe a few megabytes?). I will remove it from this PR and ask TheForge for further info (perhaps it's a lot more RAM than we think?). It may also be a better idea to destroy shader modules after each PSO creation (unless they're shared), instead of doing it at higher level.~ Done for now. We'll revisit this later.

DarioSamo commented on servers/rendering/rendering_device_commons.h:
```diff
@@ -271,6 +271,28 @@ class RenderingDeviceCommons : public Object {
 		DATA_FORMAT_MAX,
 	};
 
+	enum BreadcrumbMarker {
+		NONE = 0,
+		// Environment
+		REFLECTION_PROBES,
+		SKY_PASS,
+		// Light mapping
+		LIGHTMAPPER_PASS,
+		// Shadows
+		SHADOW_PASS_DIRECTIONAL,
+		SHADOW_PASS_CUBE,
+		// Geometry passes
+		OPAQUE_PASS,
+		ALPHA_PASS,
+		TRANSPARENT_PASS,
+		// Screen effects
+		POST_PROCESSING_PASS,
+		BLIT_PASS,
+		UI_PASS,
+		// Other
+		DEBUG_PASS
+	};
+
```
I feel the responsibility of RenderingDevice should stay as abstract as possible in this case and not have knowledge of how the rendering pipeline works. Perhaps we can keep the breadcrumbs as a typedef of just `uint32_t` and the renderers (Forward+, Mobile, or future ones) can assign the semantic meaning to the numbers?

darksylinc replied:
Breadcrumbs don't really add knowledge.

They're just tagging an ID. The idea is that if the app crashes we look at the report and see "what was being executed?" oh breadcrumb says ID = 2, which corresponds to... let me see... SKY_PASS.

It's up to the rendering pipelines to _actually_ use those IDs.

In fact some parts of the code do the following:

```
breadcrumb = LIGHTMAPPER_PASS | (pass_id << 16)
```

In other words, it doesn't give knowledge about the upper levels; it's just a way of standardizing some common tags that are more likely to be used.

darksylinc replied:
But wouldn't that mean that:

1. Mobile and Forward+ would have their own copies of the enum, which can get out of sync. The main issue about that is that it can cause silly mistakes when debugging that can waste hours (i.e. user got id = 7, dev looked it up what that means, but saw the wrong version).
2. Users who want to use the `RenderingDevice` directly have no information about what that id is for, and even if they do, they would not adhere to any convention.

Btw I looked again and turns out only 16 bits of each breadcrumb argument are used:

```cpp
uint32_t breadcrumb = (p_phase << 16) | (p_breadcrumb_data & ((1 << 16) - 1));
```

Perhaps we could change this argument signature to...?

// original:
void draw_list_begin(RID p_framebuffer, InitialAction p_initial_color_action, FinalAction p_final_color_action, InitialAction p_initial_depth_action, FinalAction p_final_depth_action, const Vector<Color> &p_clear_color_values, float p_clear_depth, uint32_t p_clear_stencil, const Rect2 &p_region, RDD::BreadcrumbMarker p_phase, uint32_t p_breadcrumb_data) {}

// My proposal:
enum BreadcrumbMarker {
   // Stays the same.
}

uint32_t breadcrumb( BreadcrumbMarker p_phase, uint16_t p_data ) {
  return (p_phase << 16) | (p_data & 0xFFFF)
}

void draw_list_begin(RID p_framebuffer, InitialAction p_initial_color_action, FinalAction p_final_color_action, InitialAction p_initial_depth_action, FinalAction p_final_depth_action, const Vector<Color> &p_clear_color_values, float p_clear_depth, uint32_t p_clear_stencil, const Rect2 &p_region, uint32_t p_breadcrumb_data) {}
```

This way the BreadcrumbMarker enum is still visible to everyone, but it is made clear the underlying datatype is an arbitrary uint32_t.

clayjohn replied:
Looking back on this. I like your suggested approach. I really don't love the RDD having high level concepts like `POST_PROCESSING_PASS` or `SHADOW_PASS_CUBE`, but at the same time, to make this feature useful we want the debug utility to print out the human readable name for the pass and not force users to look it up. You are right that looking up an ID would be a huge annoyance and significantly reduce the value of breadcrumbs.

DarioSamo commented on servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:
```diff
@@ -2166,9 +2169,7 @@ void RenderForwardMobile::_render_list_template(RenderingDevice::DrawListID p_dr
-		PipelineCacheRD *pipeline = nullptr;
-
-		pipeline = &shader->pipelines[cull_variant][primitive][shader_version];
+		PipelineCacheRD *pipeline = &shader->pipelines[cull_variant][primitive][shader_version];
```
Just as a heads up, we will have a merge conflict on this case as this code will be replaced in #90400. Seeing as it's for aesthetic reasons I'd forego this change and the removal of the comment below.

darksylinc replied:
Ugh. This is exactly why I included it (so someone could notice it, it wasn't an aesthetic change but more like bait).

There are more upcoming changes that will affect `RenderForwardMobile::_render_list_template`, as there is an optimization that uses Spec Constants for key in sorting the draw calls for better draw call batching.

darksylinc replied:
I will probably have to ask for your help when the time comes.

bruvzg commented on core/os/memory.cpp:
```diff
@@ -65,6 +66,32 @@ SafeNumeric<uint64_t> Memory::max_usage;
 
+void *Memory::alloc_aligned_static(size_t p_bytes, size_t p_alignment) {
+	void *p1, *p2;
+	if ((p1 = (void *)malloc(p_bytes + p_alignment - 1 + sizeof(uint32_t))) == NULL)
+		return NULL;
```
```suggestion
	if ((p1 = (void *)malloc(p_bytes + p_alignment - 1 + sizeof(uint32_t))) == nullptr) {
		return nullptr;
	}
```

darksylinc replied:
Done.

bruvzg commented on core/os/memory.cpp:
```diff
@@ -65,6 +66,32 @@ SafeNumeric<uint64_t> Memory::max_usage;
+}
+
+void *Memory::realloc_aligned_static(void *p_memory, size_t p_bytes, size_t p_prev_bytes, size_t p_alignment) {
+	if (p_memory == NULL)
+		return alloc_aligned_static(p_bytes, p_alignment);
```
```suggestion
	if (p_memory == nullptr) {
		return alloc_aligned_static(p_bytes, p_alignment);
	}
```

darksylinc replied:
Done.

bruvzg commented on servers/rendering/rendering_device.cpp:
```diff
@@ -5595,6 +5664,19 @@ void RenderingDevice::_bind_methods() {
 
+	ClassDB::bind_method(D_METHOD("get_perf_report"), &RenderingDevice::get_perf_report);
+
+	ClassDB::bind_method(D_METHOD("get_tracked_object_name"), &RenderingDevice::get_tracked_object_name);
```
```suggestion
	ClassDB::bind_method(D_METHOD("get_tracked_object_name", "type_index"), &RenderingDevice::get_tracked_object_name);
```

darksylinc replied:
Done.

bruvzg commented on servers/rendering/rendering_device.cpp:
```diff
@@ -5595,6 +5664,19 @@ void RenderingDevice::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("get_tracked_object_type_count"), &RenderingDevice::get_tracked_object_type_count);
+	ClassDB::bind_method(D_METHOD("get_driver_total_memory"), &RenderingDevice::get_driver_total_memory);
+	ClassDB::bind_method(D_METHOD("get_driver_allocation_count"), &RenderingDevice::get_driver_allocation_count);
+	ClassDB::bind_method(D_METHOD("get_driver_memory_by_object_type"), &RenderingDevice::get_driver_memory_by_object_type);
+	ClassDB::bind_method(D_METHOD("get_driver_allocs_by_object_type"), &RenderingDevice::get_driver_allocs_by_object_type);
```
```suggestion
	ClassDB::bind_method(D_METHOD("get_driver_memory_by_object_type", "type"), &RenderingDevice::get_driver_memory_by_object_type);
	ClassDB::bind_method(D_METHOD("get_driver_allocs_by_object_type", "type"), &RenderingDevice::get_driver_allocs_by_object_type);
```

darksylinc replied:
Done.

bruvzg commented on servers/rendering/rendering_device.cpp:
```diff
@@ -5595,6 +5664,19 @@ void RenderingDevice::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("get_driver_allocs_by_object_type"), &RenderingDevice::get_driver_allocs_by_object_type);
+	ClassDB::bind_method(D_METHOD("get_device_total_memory"), &RenderingDevice::get_device_total_memory);
+	ClassDB::bind_method(D_METHOD("get_device_allocation_count"), &RenderingDevice::get_device_allocation_count);
+	ClassDB::bind_method(D_METHOD("get_device_memory_by_object_type"), &RenderingDevice::get_device_memory_by_object_type);
+	ClassDB::bind_method(D_METHOD("get_device_allocs_by_object_type"), &RenderingDevice::get_device_allocs_by_object_type);
```
```suggestion
	ClassDB::bind_method(D_METHOD("get_device_memory_by_object_type", "type"), &RenderingDevice::get_device_memory_by_object_type);
	ClassDB::bind_method(D_METHOD("get_device_allocs_by_object_type", "type"), &RenderingDevice::get_device_allocs_by_object_type);
```

Binds should include argument names.

darksylinc replied:
Done.

bruvzg commented on drivers/vulkan/rendering_context_driver_vulkan.cpp:
```diff
@@ -40,21 +40,334 @@
+
+			uint8_t *ret = reinterpret_cast<uint8_t *>(Memory::alloc_aligned_static(size + alignment, alignment));
+			if (ret == nullptr) {
+				return NULL;
```
```suggestion
				return nullptr;
```

darksylinc replied:
Done.

bruvzg commented on drivers/vulkan/rendering_context_driver_vulkan.cpp:
```diff
@@ -40,21 +40,334 @@
+
+			uint8_t *ret = reinterpret_cast<uint8_t *>(Memory::realloc_aligned_static(header, size + alignment, header->size + alignment, alignment));
+			if (ret == nullptr) {
+				return NULL;
```
```suggestion
				return nullptr;
```

darksylinc replied:
Done.

bruvzg commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -4691,9 +4892,157 @@ void RenderingDeviceDriverVulkan::command_begin_label(CommandBufferID p_cmd_buff
+			const VkDeviceAddress lower = (addr_info->reportedAddress & ~(addr_info->addressPrecision - 1));
+			const VkDeviceAddress upper = (addr_info->reportedAddress | (addr_info->addressPrecision - 1));
+			err_msg += "Info " + itos(ad);
+			err_msg += "   Type            : " + String(addressTypeNames[addr_info->addressType]);
+			err_msg += "   Reported address: " + itos(addr_info->reportedAddress);
+			err_msg += "   Lower address   : " + itos(lower);
+			err_msg += "   Upper address   : " + itos(upper);
+			err_msg += "   Precision       : " + itos(addr_info->addressPrecision);
```
This probably should have new lines add at the end of each statement.

clayjohn replied:
I agree. I added a suggestion to in my review below

bruvzg commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -4691,9 +4892,157 @@ void RenderingDeviceDriverVulkan::command_begin_label(CommandBufferID p_cmd_buff
+			"INSTRUCTION_POINTER_INVALID",
+			"INSTRUCTION_POINTER_FAULT",
+		};
+		err_msg += "Address infosn\n";
```
"infosn"?

bruvzg commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -4691,9 +4892,157 @@ void RenderingDeviceDriverVulkan::command_begin_label(CommandBufferID p_cmd_buff
+	fault_info.sType = VK_STRUCTURE_TYPE_DEVICE_FAULT_INFO_EXT;
+	fault_info.pVendorInfos = fault_counts.vendorInfoCount
+			? (VkDeviceFaultVendorInfoEXT *)memalloc(fault_counts.vendorInfoCount * sizeof(VkDeviceFaultVendorInfoEXT))
+			: NULL;
```
```suggestion
			: nullptr;
```

bruvzg commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -4691,9 +4892,157 @@ void RenderingDeviceDriverVulkan::command_begin_label(CommandBufferID p_cmd_buff
+	fault_info.pAddressInfos =
+			fault_counts.addressInfoCount
+			? (VkDeviceFaultAddressInfoEXT *)memalloc(fault_counts.addressInfoCount * sizeof(VkDeviceFaultAddressInfoEXT))
+			: NULL;
```
```suggestion
			: nullptr;
```

bruvzg commented on core/os/memory.h:
```diff
@@ -62,6 +62,10 @@ class Memory {
 	static void *realloc_static(void *p_memory, size_t p_bytes, bool p_pad_align = false);
 	static void free_static(void *p_ptr, bool p_pad_align = false);
 
+	static void *alloc_aligned_static(size_t p_bytes, size_t p_alignment);
```
Would be nice to have some comment with description of what is exactly allocated and how metadata is stored/aligned.

clayjohn replied:
Agreed

darksylinc replied:
Done.

bruvzg commented on servers/rendering/rendering_device_driver.h:
```diff
@@ -468,6 +468,7 @@ class RenderingDeviceDriver : public RenderingDeviceCommons {
 	// Only meaningful if API_TRAIT_SHADER_CHANGE_INVALIDATION is SHADER_CHANGE_INVALIDATION_ALL_OR_NONE_ACCORDING_TO_LAYOUT_HASH.
 	virtual uint32_t shader_get_layout_hash(ShaderID p_shader) { return 0; }
 	virtual void shader_free(ShaderID p_shader) = 0;
+	virtual void shader_destroy_modules(ShaderID p_shader) = 0;
```
Imp mentation is missing from DX12 driver (or it can be changed to non-abstract dummy here).

bruvzg commented on servers/rendering/rendering_device_driver.h:
```diff
@@ -701,6 +702,11 @@ class RenderingDeviceDriver : public RenderingDeviceCommons {
+	/****************/
+	/**** DEBUG *****/
+	/****************/
+	virtual void command_insert_breadcrumb(CommandBufferID p_cmd_buffer, uint32_t p_data) = 0;
```
Same here.

bruvzg commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -4691,9 +4892,157 @@ void RenderingDeviceDriverVulkan::command_begin_label(CommandBufferID p_cmd_buff
+			? (VkDeviceFaultAddressInfoEXT *)memalloc(fault_counts.addressInfoCount * sizeof(VkDeviceFaultAddressInfoEXT))
+			: NULL;
+	fault_counts.vendorBinarySize = 0;
+	vkres = vkGetDeviceFaultInfoEXT(vk_device, &fault_counts, &fault_info);
```
`vkGetDeviceFaultInfoEXT` seems to be 1.3 extensions, function pointer should be loaded dynamically.

darksylinc replied:
Done.

AThousandShips commented on servers/rendering/rendering_device.cpp:
```diff
@@ -2478,6 +2495,23 @@ RID RenderingDevice::shader_create_from_bytecode(const Vector<uint8_t> &p_shader
+void RenderingDevice::_destroy_all_shader_modules() {
+	List<RID> remaining;
+	shader_owner.get_owned_list(&remaining);
+	if (remaining.size()) {
```
Unnecessary double check

darksylinc replied:
Done.

darksylinc commented on servers/rendering/rendering_device.h:
```diff
@@ -814,12 +820,15 @@ class RenderingDevice : public RenderingDeviceCommons {
 	RID shader_create_from_spirv(const Vector<ShaderStageSPIRVData> &p_spirv, const String &p_shader_name = "");
 	RID shader_create_from_bytecode(const Vector<uint8_t> &p_shader_binary, RID p_placeholder = RID());
 	RID shader_create_placeholder();
+	void shader_destroy_modules(RID p_shaderRID);
```
Done.

darksylinc commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -4691,9 +4892,157 @@ void RenderingDeviceDriverVulkan::command_begin_label(CommandBufferID p_cmd_buff
+		return;
+	}
+
+	String err_msg = "";
```
Done.

darksylinc commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -4691,9 +4892,157 @@ void RenderingDeviceDriverVulkan::command_begin_label(CommandBufferID p_cmd_buff
+	vmaUnmapMemory(allocator, ((BufferInfo *)breadcrumb_buffer.id)->allocation.handle);
+	uint32_t phase = last_breadcrumb >> 16;
+	uint32_t user_data = last_breadcrumb & ((1 << 16) - 1);
+	String errorMsg = "Last known breadcrumb: ";
```
Done.

darksylinc commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -4496,6 +4680,8 @@ RDD::PipelineID RenderingDeviceDriverVulkan::render_pipeline_create(
 	pipeline_create_info.stageCount = shader_info->vk_stages_create_info.size();
 
+	ERR_FAIL_COND_V_MSG(pipeline_create_info.stageCount == 0, PipelineID(),
+			"Cannot create pipeline without shader module, please make sure shader modules are destroyed only after all associated pipelines are created");
```
Done.

darksylinc commented on servers/rendering/rendering_device.cpp:
```diff
@@ -5288,6 +5326,37 @@ uint64_t RenderingDevice::get_driver_resource(DriverResource p_resource, RID p_r
+String RenderingDevice::get_tracked_object_name(uint32_t p_type_index) const {
+	return context->get_tracked_object_name(p_type_index);
+}
+uint64_t RenderingDevice::get_tracked_object_type_count() const {
```
Done.

darksylinc commented on servers/rendering/rendering_context_driver.cpp:
```diff
@@ -83,3 +83,34 @@ void RenderingContextDriver::window_destroy(DisplayServer::WindowID p_window) {
+const char *RenderingContextDriver::get_tracked_object_name(uint32_t p_type_index) const {
+	return "Tracking Unsupported by API";
+}
+uint64_t RenderingContextDriver::get_tracked_object_type_count() const {
```
Done.

darksylinc commented on servers/rendering/rendering_device.cpp:
```diff
@@ -5288,6 +5326,37 @@ uint64_t RenderingDevice::get_driver_resource(DriverResource p_resource, RID p_r
+	return context->get_device_memory_by_object_type(type);
+}
+uint64_t RenderingDevice::get_device_allocs_by_object_type(uint32_t type) const {
+	return context->get_device_allocs_by_object_type(type);
```
Done.

darksylinc commented on drivers/vulkan/rendering_context_driver_vulkan.cpp:
```diff
@@ -40,21 +40,334 @@
+uint64_t RenderingContextDriverVulkan::get_device_total_memory() const {
+	return memory_report_total_memory.get();
+}
+uint64_t RenderingContextDriverVulkan::get_device_allocation_count() const {
```
Done.

darksylinc commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -4691,9 +4892,157 @@ void RenderingDeviceDriverVulkan::command_begin_label(CommandBufferID p_cmd_buff
+	if (fault_info.pVendorInfos)
+		memfree(fault_info.pVendorInfos);
+	if (fault_info.pAddressInfos)
+		memfree(fault_info.pAddressInfos);
```
Done.

darksylinc commented on doc/classes/RenderingDevice.xml:
```diff
@@ -527,6 +591,33 @@
+				- PIPELINE_CACHE
+				- SWAPCHAIN_KHR
+				- COMMAND_POOL
+				Thus if e.g. get_tracked_object_name(5) returns "COMMAND_POOL", then get_device_memory_by_object_type(5) returns the bytes used by the GPU for command pools.
```
Done. You may wanna check this one again though.

darksylinc commented on doc/classes/RenderingDevice.xml:
```diff
@@ -527,6 +591,33 @@
+			<param index="0" name="type" type="int" />
+			<description>
+				Returns the name of the type of object for the given [param type]. This value must be in range \[0; get_tracked_object_type_count). If get_tracked_object_type_count is 0, then type argument is ignored and always returns the same string.
+				The return value is important because it gives meaning to the types passed to get_driver_memory_by_object_type, get_driver_allocs_by_object_type, get_device_memory_by_object_type and get_device_allocs_by_object_type. Examples of strings it can return (not exhaustive):
```
Done. You may wanna check this one again though.

AThousandShips replied:
This one is still valid and should be applied, all methods should be linked I'd say

darksylinc commented:
I've made an update. Look at the "3 outstanding things remain:" in the top post where I ask for a bit of help :smile:

AThousandShips commented on doc/classes/RenderingDevice.xml:
```diff
@@ -487,6 +493,31 @@
+			<param index="0" name="type" type="int" />
+			<description>
+				Same as get_device_allocation_count but filtered for a given object type.
+				The type argument must be in range [code]【0; get_tracked_object_type_count)[/code]. If get_tracked_object_type_count is 0, then type argument is ignored and always returns 0.
```
```suggestion
				The type argument must be in range [code][0, get_tracked_object_type_count)[/code]. If [method get_tracked_object_type_count] is 0, then type argument is ignored and always returns 0.
```

darksylinc replied:
**Small warning:** I cannot change `【0; get_tracked_object_type_count)` to `[0; get_tracked_object_type_count)` because it fails pre commit checks with:

```
ERROR: RenderingDevice.xml: Tag depth mismatch: too many (or too few) open/close tags in method "get_driver_allocs_by_object_type" description.
```

I tried:
 - \\[
 - \`[\`
 - \`\\[\`

The error appears whenever the description has includes some other special tag like `[method]` or `[code]`.

If you have a solution I'll gladly use it; but meanwhile I had to workaround it by using the Unicode similar character 【

AThousandShips replied:
Ah then it's an escape, but I'd say since it's an integer here you should instead do:
```suggestion
				The type argument must be in range [code][0, get_tracked_object_type_count - 1][/code]. If [method get_tracked_object_type_count] is 0, then type argument is ignored and always returns 0.
```

darksylinc replied:
What a simple solution.
I hate it. And I love it.

While I prefer `[0, get_tracked_object_type_count)` notation, I was unhappy that `【` breaks text2voice for visually impaired. Your solution it is.

AThousandShips commented on doc/classes/RenderingDevice.xml:
```diff
@@ -487,6 +493,31 @@
+			<param index="0" name="type" type="int" />
+			<description>
+				Same as get_device_total_memory but filtered for a given object type.
+				The type argument must be in range [code]【0; get_tracked_object_type_count)[/code]. If get_tracked_object_type_count is 0, then type argument is ignored and always returns 0.
```
```suggestion
				The type argument must be in range [code][0, get_tracked_object_type_count)[/code]. If [method get_tracked_object_type_count] is 0, then type argument is ignored and always returns 0.
```

AThousandShips commented on doc/classes/RenderingDevice.xml:
```diff
@@ -499,12 +530,44 @@
+			<param index="0" name="type" type="int" />
+			<description>
+				Same as get_driver_allocation_count but filtered for a given object type.
+				The type argument must be in range [code]【0; get_tracked_object_type_count)[/code]. If get_tracked_object_type_count is 0, then type argument is ignored and always returns 0.
```
```suggestion
				The type argument must be in range [code][0, get_tracked_object_type_count)[/code]. If [method get_tracked_object_type_count] is 0, then type argument is ignored and always returns 0.
```

AThousandShips commented on doc/classes/RenderingDevice.xml:
```diff
@@ -499,12 +530,44 @@
+			<param index="0" name="type" type="int" />
+			<description>
+				Same as get_driver_total_memory but filtered for a given object type.
+				The type argument must be in range [code]【0; get_tracked_object_type_count)[/code]. If get_tracked_object_type_count is 0, then type argument is ignored and always returns 0.
```
```suggestion
				The type argument must be in range [code][0, get_tracked_object_type_count)[/code]. If [method get_tracked_object_type_count] is 0, then type argument is ignored and always returns 0.
```

AThousandShips commented on doc/classes/RenderingDevice.xml:
```diff
@@ -527,6 +597,33 @@
+			<return type="String" />
+			<param index="0" name="type_index" type="int" />
+			<description>
+				Returns the name of the type of object for the given [param type_index]. This value must be in range [code]【0; get_tracked_object_type_count)[/code]. If get_tracked_object_type_count is 0, then type argument is ignored and always returns the same string.
```
```suggestion
				Returns the name of the type of object for the given [param type_index]. This value must be in range [code][0, get_tracked_object_type_count)[/code]. If [method get_tracked_object_type_count] is 0, then type argument is ignored and always returns the same string.
```

AThousandShips commented on servers/rendering/rendering_device.cpp:
```diff
@@ -513,9 +515,24 @@ Error RenderingDevice::buffer_update(RID p_buffer, uint32_t p_offset, uint32_t p
+}
+
+void RenderingDevice::update_perf_report() {
+	perf_report_text = "";
+	perf_report_text += " gpu:" + String::num_int64(gpu_copy_count);
```
```suggestion
	perf_report_text = " gpu:" + String::num_int64(gpu_copy_count);
```

darksylinc commented:
All checks passed!

clayjohn commented on drivers/vulkan/rendering_device_driver_vulkan.cpp:
```diff
@@ -4693,9 +4899,165 @@ void RenderingDeviceDriverVulkan::command_begin_label(CommandBufferID p_cmd_buff
+			// From https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkDeviceFaultAddressInfoEXT.html
+			const VkDeviceAddress lower = (addr_info->reportedAddress & ~(addr_info->addressPrecision - 1));
+			const VkDeviceAddress upper = (addr_info->reportedAddress | (addr_info->addressPrecision - 1));
+			err_msg += "Info " + itos(ad);
+			err_msg += "   Type            : " + String(addressTypeNames[addr_info->addressType]);
+			err_msg += "   Reported address: " + itos(addr_info->reportedAddress);
+			err_msg += "   Lower address   : " + itos(lower);
+			err_msg += "   Upper address   : " + itos(upper);
+			err_msg += "   Precision       : " + itos(addr_info->addressPrecision);
```
```suggestion
			err_msg += "Info " + itos(ad) + "\n";
			err_msg += "   Type            : " + String(addressTypeNames[addr_info->addressType]) + "\n";
			err_msg += "   Reported address: " + itos(addr_info->reportedAddress) + "\n";
			err_msg += "   Lower address   : " + itos(lower) + "\n";
			err_msg += "   Upper address   : " + itos(upper) + "\n";
			err_msg += "   Precision       : " + itos(addr_info->addressPrecision) + "\n";
```
Following up on bruvzg's review from earlier.

darksylinc replied:
Done.

I went for putting the newline in the beginning of the hardcoded string because it has multiple advantages.

i.e.
```cpp
err_msg += "\n    Message : " + itos(lower);
```

instead of:

```cpp
err_msg += "    Message : " + itos(lower) + "\n";
```

clayjohn reviewed (approved):
Looks good to me to merge for Beta 1 pending the small changes suggested by Bruvzg. 

I would like to get this in to Beta 1 as it will help massively for debugging Vulkan issues. We fixed a lot of the significant bugs in 4.3 already, the rest are going to be nearly impossible to track down without this additional info. So IMO, its best to get this in now so that users of 4.3 that still run into Vulkan bugs can give us a sensible report by running a debug build

AThousandShips commented on doc/classes/RenderingDevice.xml:
```diff
@@ -527,6 +597,33 @@
+			<param index="0" name="type_index" type="int" />
+			<description>
+				Returns the name of the type of object for the given [param type_index]. This value must be in range [code][0; get_tracked_object_type_count - 1][/code]. If [method get_tracked_object_type_count] is 0, then type argument is ignored and always returns the same string.
+				The return value is important because it gives meaning to the types passed to [method get_driver_memory_by_object_type], [method get_driver_allocs_by_object_type], [method get_device_memory_by_object_type] and [method get_device_allocs_by_object_type]. Examples of strings it can return (not exhaustive):
```
```suggestion
				The return value is important because it gives meaning to the types passed to [method get_driver_memory_by_object_type], [method get_driver_allocs_by_object_type], [method get_device_memory_by_object_type], and [method get_device_allocs_by_object_type]. Examples of strings it can return (not exhaustive):
```
Please use [Oxford comma](https://docs.godotengine.org/en/latest/contributing/documentation/docs_writing_guidelines.html#use-the-oxford-comma-to-enumerate-anything).

AThousandShips commented on servers/rendering/rendering_device.cpp:
```diff
@@ -5591,7 +5650,7 @@ void RenderingDevice::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("draw_list_begin_for_screen", "screen", "clear_color"), &RenderingDevice::draw_list_begin_for_screen, DEFVAL(DisplayServer::MAIN_WINDOW_ID), DEFVAL(Color()));
 
-	ClassDB::bind_method(D_METHOD("draw_list_begin", "framebuffer", "initial_color_action", "final_color_action", "initial_depth_action", "final_depth_action", "clear_color_values", "clear_depth", "clear_stencil", "region"), &RenderingDevice::draw_list_begin, DEFVAL(Vector<Color>()), DEFVAL(1.0), DEFVAL(0), DEFVAL(Rect2()));
+	ClassDB::bind_method(D_METHOD("draw_list_begin", "framebuffer", "initial_color_action", "final_color_action", "initial_depth_action", "final_depth_action", "clear_color_values", "clear_depth", "clear_stencil", "region", "breadcrumb"), &RenderingDevice::draw_list_begin, DEFVAL(Vector<Color>()), DEFVAL(1.0), DEFVAL(0), DEFVAL(Rect2()), DEFVAL(0));
```
This method needs a compatibility bind, see https://github.com/godotengine/godot-docs/pull/9158

AThousandShips reviewed (dismissed):
See above with compat

AThousandShips commented on misc/extension_api_validation/4.2-stable.expected:
```diff
@@ -357,3 +356,14 @@ Validate extension JSON: Error: Field 'classes/RenderingServer/methods/canvas_it
+--------
+Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
+Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': default_value changed value in new API, from "Array[RID]([])" to "0".
+Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "typedarray::RID" to "int".
```
Only one of these should be kept I think, the type part

darksylinc replied:
The last two are complaints from 4.2 -> 4.3
But the last one is a complaint from 4.1 (or 4.0?) -> 4.3

I was a bit confused by it because the same issue should've happened when going 4.1 -> 4.2

akien-mga replied:
Hm there seems to be a clash with #84976, which is a 4.2->4.3 change that _removed_ the 9th argument to `draw_list_begin`.

Now this PR adds a new 9th argument with a different type.

I think it's the first time we have this case, not sure if the script we use to check for compatibility issues is able to handle that. CC @dsnopek

dsnopek replied:
Since both changes were in 4.3, I would expect it to treat this like the 9th argument changed type, completely ignoring the previous removal of the argument.

The weird thing is these two lines:

```
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "typedarray::RID" to "int".
```

I'm not sure how it can change from both `Array` and `typedarray::RID`.

Looking at the [reference `extension_api.json`](https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.2-stable/gdextension/extension_api.json), it seems like `typedarray::RID` is the correct one.

What happens if you remove this line from the `4.2-stable.expected`:

```
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
```

darksylinc replied:
If I remove the first entry and only leave:

```
GH-90993
--------
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': default_value changed value in new API, from "Array[RID]([])" to "0".
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "typedarray::RID" to "int".

draw_list_begin added a new optional debug argument called breadcrumb.
There used to be an RID argument as arg #9 which was removed in GH-84976 but now we're adding a new one in the same location.
No compatibility method was registered because the one created for GH-84976 is already handling this case.
```

Then the CI still fails. The output is:

```
Godot Engine v4.3.beta.custom_build.2446b419d (2024-05-18 16:21:37 UTC) - https://godotengine.org

2024-05-18 17:07:59 URL:https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.1-stable/gdextension/extension_api.json [5318978/5318978] -> "/tmp/tmp.exex1R1Xhb" [1]

ERROR: New API lacks base array: enums
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1362)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/draw_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_begin': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
Compatibility to 4.1-stable is broken in the following ways:
Validate extension JSON: API was removed: classes/OpenXRAction
Validate extension JSON: API was removed: classes/OpenXRActionMap
Validate extension JSON: API was removed: classes/OpenXRActionSet
Validate extension JSON: API was removed: classes/OpenXRHand
Validate extension JSON: API was removed: classes/OpenXRInteractionProfile
Validate extension JSON: API was removed: classes/OpenXRInterface
Validate extension JSON: API was removed: classes/OpenXRIPBinding
Validate extension JSON: API was removed: classes/WebXRInterface
Godot Engine v4.3.beta.custom_build.2446b419d (2024-05-18 16:21:37 UTC) - https://godotengine.org

2024-05-18 17:12:27 URL:https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.0-stable/gdextension/extension_api.json [5175722/5175722] -> "/tmp/tmp.qkBYdofWdn" [1]


ERROR: New API lacks base array: enums
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1362)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/draw_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_begin': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
Compatibility to 4.0-stable is broken in the following ways:
Validate extension JSON: API was removed: classes/OpenXRAction
Validate extension JSON: API was removed: classes/OpenXRActionMap
Validate extension JSON: API was removed: classes/OpenXRActionSet
Validate extension JSON: API was removed: classes/OpenXRHand
Validate extension JSON: API was removed: classes/OpenXRInteractionProfile
Validate extension JSON: API was removed: classes/OpenXRInterface
Validate extension JSON: API was removed: classes/OpenXRIPBinding
Validate extension JSON: API was removed: classes/WebXRInterface
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
Godot Engine v4.3.beta.custom_build.2446b419d (2024-05-18 16:21:37 UTC) - https://godotengine.org

2024-05-18 17:12:29 URL:https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.2-stable/gdextension/extension_api.json [5501700/5501700] -> "/tmp/tmp.vlgKzMe49U" [1]

ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/draw_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_begin': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
The following validation errors no longer occur (compared to 4.2-stable):
Validate extension JSON: Error: Field 'classes/WebXRInterface/methods/get_input_source_tracker/return_value': type changed value in new API, from "XRPositionalTracker" to "XRControllerTracker".
Compatibility to 4.2-stable is broken in the following ways:
Validate extension JSON: API was removed: classes/OpenXRAction
Validate extension JSON: API was removed: classes/OpenXRActionMap
Validate extension JSON: API was removed: classes/OpenXRActionSet
Validate extension JSON: API was removed: classes/OpenXRAPIExtension
Validate extension JSON: API was removed: classes/OpenXRExtensionWrapperExtension
Validate extension JSON: API was removed: classes/OpenXRHand
Validate extension JSON: API was removed: classes/OpenXRInteractionProfile
Validate extension JSON: API was removed: classes/OpenXRInteractionProfileMetadata
Validate extension JSON: API was removed: classes/OpenXRInterface
Validate extension JSON: API was removed: classes/OpenXRIPBinding
Validate extension JSON: API was removed: classes/WebXRInterface
```

**[AS I SAID](https://github.com/godotengine/godot/pull/90993#discussion_r1605091726)** this error appears in validation from [4.0](https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.0-stable/gdextension/extension_api.json) -> 4.3

If I restore the
`Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".` then 4.0 -> 4.3 passes correctly, but [4.2](https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.2-stable/gdextension/extension_api.json) -> 4.3 will say:

```
The following validation errors no longer occur (compared to 4.2-stable):
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
```

darksylinc replied:
In [godot-4.0-stable/gdextension/extension_api.json](https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.0-stable/gdextension/extension_api.json):

```json
{
	"name": "storage_textures",
	"type": "Array",
	"default_value": "Array[RID]([])"
}
```

In [godot-4.1-stable/gdextension/extension_api.json](https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.1-stable/gdextension/extension_api.json) & [godot-4.2-stable/gdextension/extension_api.json](https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.2-stable/gdextension/extension_api.json):

```json
{
	"name": "storage_textures",
	"type": "typedarray::RID",
	"default_value": "Array[RID]([])"
}
```

akien-mga replied:
I'll look into it myself, I think Matias' time is better spent in rendering issues than in debugging our CI API compat testing framework.

akien-mga replied:
IMO our validation scripts are buggy and need to be fixed to properly handle APIs changing multiple times across versions.

But for now this should be an ok workaround to silence both the error if not adding the line, and the false positive warning when adding it in the `4.2` file.

```diff
diff --git a/misc/extension_api_validation/4.0-stable_4.1-stable.expected b/misc/extension_api_validation/4.0-stable_4.1-stable.expected
index 5c3bf07fb2..7388c90e2e 100644
--- a/misc/extension_api_validation/4.0-stable_4.1-stable.expected
+++ b/misc/extension_api_validation/4.0-stable_4.1-stable.expected
@@ -349,3 +349,11 @@ Validate extension JSON: Error: Hash changed for 'classes/EditorUndoRedoManager/
 Validate extension JSON: Error: Hash changed for 'classes/UndoRedo/methods/create_action', from 0AEC1BFC to E87757EB. This means that the function has changed and no compatibility function was provided.
 
 Added a optional parameters with default values. No adjustments should be necessary.
+
+
+GH-90993
+--------
+Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
+
+The change to `int` actually happened from 4.2 to 4.3, but our current validation script doesn't handle well the case where an argument changed type twice in different versions.
+This argument already changed in 4.1 in GH-76418, so the script gets confused.
diff --git a/misc/extension_api_validation/4.2-stable.expected b/misc/extension_api_validation/4.2-stable.expected
index 54f95cde58..cf80419cf9 100644
--- a/misc/extension_api_validation/4.2-stable.expected
+++ b/misc/extension_api_validation/4.2-stable.expected
@@ -367,7 +367,6 @@ Optional arguments added. Compatibility methods registered.
 
 GH-90993
 --------
-Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
 Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': default_value changed value in new API, from "Array[RID]([])" to "0".
 Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "typedarray::RID" to "int".
 
```

darksylinc replied:
Done!!!

I've applied your patch, rebased to latest master, and pushed.

Now all CI passes!

darksylinc commented on platform/linuxbsd/x11/rendering_context_driver_vulkan_x11.cpp:
```diff
@@ -51,7 +51,7 @@ RenderingContextDriver::SurfaceID RenderingContextDriverVulkanX11::surface_creat
 
 	VkSurfaceKHR vk_surface = VK_NULL_HANDLE;
-	VkResult err = vkCreateXlibSurfaceKHR(instance_get(), &create_info, nullptr, &vk_surface);
+	VkResult err = vkCreateXlibSurfaceKHR(instance_get(), &create_info, get_allocation_callbacks(VK_OBJECT_TYPE_SURFACE_KHR), &vk_surface);
```
Ooof! Well spotted!

macOS was also missing it.
Fixed!

RandomShaper commented on core/os/memory.cpp:
```diff
@@ -65,6 +66,38 @@ SafeNumeric<uint64_t> Memory::max_usage;
+	return p2;
+}
+
+void *Memory::realloc_aligned_static(void *p_memory, size_t p_bytes, size_t p_prev_bytes, size_t p_alignment) {
+	if (p_memory == nullptr) {
+		return alloc_aligned_static(p_bytes, p_alignment);
+	}
+
+	void *ret = alloc_aligned_static(p_bytes, p_alignment);
+	memcpy(ret, p_memory, p_prev_bytes);
+	free_aligned_static(p_memory);
+	return ret;
+}
```
I'd say this has to be on `realloc_static()`, to avoid pessimizing for the case where the address changes.

darksylinc replied:
Could you elaborate? Because I do not understand what you are asking.

RandomShaper replied:
My point is that a reallocation doesn't necessarilly involve a new address. When the allocator can just resize the block and give you the same address, you can just avoid any further operations. In contrast, this implementation will never benefit from that and will unconditionally pay for the overhead of dealing with a second allocation and, more important, that of copying the data.

darksylinc replied:
aligned_realloc is a minefield and relatively [hard to implement](https://stackoverflow.com/questions/56619623/why-is-there-no-aligned-realloc-on-most-platforms).

Not impossible, but not straightforward (it looks straightforward, then the devil in the details appear). I would advise against it unless someone wants to write heavy unit testing to ensure it works in all edge cases.

RandomShaper replied:
Let's go with this one. I'll submit a separate PR with a more efficient implementation if I can come up with any.

darksylinc committed:
Add debug utilities for Vulkan

Features:
- Debug-only tracking of objects by type. See
get_driver_allocs_by_object_type et al.
 - Debug-only Breadcrumb info for debugging GPU crashes and device lost
 - Performance report per frame from get_perf_report
- Some VMA calls had to be modified in order to insert the necessary
memory callbacks

Functionality marked as "debug-only" is only available in debug or dev
builds.

Misc fixes:
 - Early break optimization in RenderingDevice::uniform_set_create

============================

The work was performed by collaboration of TheForge and Google. I am
merely splitting it up into smaller PRs and cleaning it up.

darksylinc commented:
I've rebased to master again now that we're in 4.4.

I already know the CI will fail because running `./misc/scripts/validate_extension_api.sh /home/matias/SlowProjects/godot/bin/godot.linuxbsd.editor.dev.x86_64.llvm` locally gave me this output:

```
./misc/scripts/validate_extension_api.sh /home/matias/SlowProjects/godot/bin/godot.linuxbsd.editor.dev.x86_64.llvm
2024-08-17 13:19:40 URL:https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.1-stable/gdextension/extension_api.json [5318978/5318978] -> "/tmp/tmp.jxZm9vNVhd" [1]
Godot Engine v4.4.dev.custom_build.1bd740d18 (2024-08-16 22:47:17 UTC) - https://godotengine.org

ERROR: New API lacks base array: enums
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1362)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/draw_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_begin': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
Compatibility to 4.1-stable is broken in the following ways:
Validate extension JSON: API was removed: classes/OpenXRAction
Validate extension JSON: API was removed: classes/OpenXRActionMap
Validate extension JSON: API was removed: classes/OpenXRActionSet
Validate extension JSON: API was removed: classes/OpenXRHand
Validate extension JSON: API was removed: classes/OpenXRInteractionProfile
Validate extension JSON: API was removed: classes/OpenXRInterface
Validate extension JSON: API was removed: classes/OpenXRIPBinding
Validate extension JSON: API was removed: classes/WebXRInterface
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': default_value changed value in new API, from "Array[RID]([])" to "0".
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "typedarray::RID" to "int".
2024-08-17 13:19:41 URL:https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.3-stable/gdextension/extension_api.json [5907082/5907082] -> "/tmp/tmp.vFi2XPNiA1" [1]
Godot Engine v4.4.dev.custom_build.1bd740d18 (2024-08-16 22:47:17 UTC) - https://godotengine.org

Compatibility to 4.3-stable is broken in the following ways:
Validate extension JSON: API was removed: classes/OpenXRAction
Validate extension JSON: API was removed: classes/OpenXRActionMap
Validate extension JSON: API was removed: classes/OpenXRActionSet
Validate extension JSON: API was removed: classes/OpenXRAPIExtension
Validate extension JSON: API was removed: classes/OpenXRCompositionLayer
Validate extension JSON: API was removed: classes/OpenXRCompositionLayerCylinder
Validate extension JSON: API was removed: classes/OpenXRCompositionLayerEquirect
Validate extension JSON: API was removed: classes/OpenXRCompositionLayerQuad
Validate extension JSON: API was removed: classes/OpenXRExtensionWrapperExtension
Validate extension JSON: API was removed: classes/OpenXRHand
Validate extension JSON: API was removed: classes/OpenXRInteractionProfile
Validate extension JSON: API was removed: classes/OpenXRInteractionProfileMetadata
Validate extension JSON: API was removed: classes/OpenXRInterface
Validate extension JSON: API was removed: classes/OpenXRIPBinding
Validate extension JSON: API was removed: classes/WebXRInterface
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments': size changed value in new API, from 9 to 10.
Validate extension JSON: Error: Hash changed for 'classes/RenderingDevice/methods/draw_list_begin', from A0225762 to FD7F8214. This means that the function has changed and no compatibility function was provided.
2024-08-17 13:19:43 URL:https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.0-stable/gdextension/extension_api.json [5175722/5175722] -> "/tmp/tmp.VCeG94saQZ" [1]
Godot Engine v4.4.dev.custom_build.1bd740d18 (2024-08-16 22:47:17 UTC) - https://godotengine.org

ERROR: New API lacks base array: enums
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1362)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/draw_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_begin': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
Compatibility to 4.0-stable is broken in the following ways:
Validate extension JSON: API was removed: classes/OpenXRAction
Validate extension JSON: API was removed: classes/OpenXRActionMap
Validate extension JSON: API was removed: classes/OpenXRActionSet
Validate extension JSON: API was removed: classes/OpenXRHand
Validate extension JSON: API was removed: classes/OpenXRInteractionProfile
Validate extension JSON: API was removed: classes/OpenXRInterface
Validate extension JSON: API was removed: classes/OpenXRIPBinding
Validate extension JSON: API was removed: classes/WebXRInterface
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': default_value changed value in new API, from "Array[RID]([])" to "0".
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
2024-08-17 13:19:45 URL:https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.2-stable/gdextension/extension_api.json [5501700/5501700] -> "/tmp/tmp.fxi2O3Si4Z" [1]
Godot Engine v4.4.dev.custom_build.1bd740d18 (2024-08-16 22:47:17 UTC) - https://godotengine.org

ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/draw_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_begin': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
Compatibility to 4.2-stable is broken in the following ways:
Validate extension JSON: API was removed: classes/OpenXRAction
Validate extension JSON: API was removed: classes/OpenXRActionMap
Validate extension JSON: API was removed: classes/OpenXRActionSet
Validate extension JSON: API was removed: classes/OpenXRAPIExtension
Validate extension JSON: API was removed: classes/OpenXRExtensionWrapperExtension
Validate extension JSON: API was removed: classes/OpenXRHand
Validate extension JSON: API was removed: classes/OpenXRInteractionProfile
Validate extension JSON: API was removed: classes/OpenXRInteractionProfileMetadata
Validate extension JSON: API was removed: classes/OpenXRInterface
Validate extension JSON: API was removed: classes/OpenXRIPBinding
Validate extension JSON: API was removed: classes/WebXRInterface
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': default_value changed value in new API, from "Array[RID]([])" to "0".
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "typedarray::RID" to "int".
```

I recognized this error when going 4.3 -> 4.4:
```
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments': size changed value in new API, from 9 to 10.
```

This is expected and should be silenced because I added `uint32_t p_breadcrumb = 0` to the following function:

```cpp
DrawListID draw_list_begin(RID p_framebuffer, InitialAction p_initial_color_action, FinalAction p_final_color_action, InitialAction p_initial_depth_action, FinalAction p_final_depth_action, const Vector<Color> &p_clear_color_values = Vector<Color>(), float p_clear_depth = 1.0, uint32_t p_clear_stencil = 0, const Rect2 &p_region = Rect2(), uint32_t p_breadcrumb = 0);
```

However this error is new and I don't know what it means or whether it's safe to ignore:

```
Validate extension JSON: Error: Hash changed for 'classes/RenderingDevice/methods/draw_list_begin', from A0225762 to FD7F8214. This means that the function has changed and no compatibility function was provided.
```

akien-mga commented:
I think you lost the changes to `misc/extension_api_validation/4.2-stable.expected` when rebasing. Pre-rebase diff: https://github.com/godotengine/godot/commit/3364acaaff8db28d24822b3fcb8b4a5693fc3c92#diff-2cc9ae1407b900dfd30f2edfeef07c9ae8cb452bac6149600a455ad1ba9f2c09

The file has been renamed to `4.2-stable_4.3-stable.expected` since 4.3-stable was released, and you need to add new stuff that breaks compat with 4.3 in the new `4.3-stable.expected` file.

darksylinc commented:
> I think you lost the changes to misc/extension_api_validation/4.2-stable.expected when rebasing. Pre-rebase diff: https://github.com/godotengine/godot/commit/3364acaaff8db28d24822b3fcb8b4a5693fc3c92#diff-2cc9ae1407b900dfd30f2edfeef07c9ae8cb452bac6149600a455ad1ba9f2c09

Oh yes, I forgot to mention that I stripped them on purpose. A refresher: The original PR ran into a validation bug because it was the first time a function (draw_list_begin) was modified twice in the same version (when going 4.2 -> 4.3) by two different PRs.

If I add to 4.3-stable.expected:

```
GH-90993
--------
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments': size changed value in new API, from 9 to 10.
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': default_value changed value in new API, from "Array[RID]([])" to "0".
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "typedarray::RID" to "int".

draw_list_begin added a new optional debug argument called breadcrumb.
There used to be an RID argument as arg #9 which was removed in GH-84976 but now we're adding a new one in the same location.
No compatibility method was registered because the one created for GH-84976 is already handling this case.
```

I get:

```
./misc/scripts/validate_extension_api.sh /home/matias/SlowProjects/godot/bin/godot.linuxbsd.editor.dev.x86_64.llvm
2024-08-17 14:37:31 URL:https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.1-stable/gdextension/extension_api.json [5318978/5318978] -> "/tmp/tmp.Lc8vSIOSbi" [1]
Godot Engine v4.4.dev.custom_build.1bd740d18 (2024-08-16 22:47:17 UTC) - https://godotengine.org

ERROR: New API lacks base array: enums
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1362)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/draw_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_begin': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
Compatibility to 4.1-stable is broken in the following ways:
Validate extension JSON: API was removed: classes/OpenXRAction
Validate extension JSON: API was removed: classes/OpenXRActionMap
Validate extension JSON: API was removed: classes/OpenXRActionSet
Validate extension JSON: API was removed: classes/OpenXRHand
Validate extension JSON: API was removed: classes/OpenXRInteractionProfile
Validate extension JSON: API was removed: classes/OpenXRInterface
Validate extension JSON: API was removed: classes/OpenXRIPBinding
Validate extension JSON: API was removed: classes/WebXRInterface
2024-08-17 14:37:33 URL:https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.3-stable/gdextension/extension_api.json [5907082/5907082] -> "/tmp/tmp.PNkdLuSnW5" [1]
Godot Engine v4.4.dev.custom_build.1bd740d18 (2024-08-16 22:47:17 UTC) - https://godotengine.org

The following validation errors no longer occur (compared to 4.3-stable):
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': default_value changed value in new API, from "Array[RID]([])" to "0".
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "Array" to "int".
Validate extension JSON: Error: Field 'classes/RenderingDevice/methods/draw_list_begin/arguments/9': type changed value in new API, from "typedarray::RID" to "int".
Compatibility to 4.3-stable is broken in the following ways:
Validate extension JSON: API was removed: classes/OpenXRAction
Validate extension JSON: API was removed: classes/OpenXRActionMap
Validate extension JSON: API was removed: classes/OpenXRActionSet
Validate extension JSON: API was removed: classes/OpenXRAPIExtension
Validate extension JSON: API was removed: classes/OpenXRCompositionLayer
Validate extension JSON: API was removed: classes/OpenXRCompositionLayerCylinder
Validate extension JSON: API was removed: classes/OpenXRCompositionLayerEquirect
Validate extension JSON: API was removed: classes/OpenXRCompositionLayerQuad
Validate extension JSON: API was removed: classes/OpenXRExtensionWrapperExtension
Validate extension JSON: API was removed: classes/OpenXRHand
Validate extension JSON: API was removed: classes/OpenXRInteractionProfile
Validate extension JSON: API was removed: classes/OpenXRInteractionProfileMetadata
Validate extension JSON: API was removed: classes/OpenXRInterface
Validate extension JSON: API was removed: classes/OpenXRIPBinding
Validate extension JSON: API was removed: classes/WebXRInterface
Validate extension JSON: Error: Hash changed for 'classes/RenderingDevice/methods/draw_list_begin', from A0225762 to FD7F8214. This means that the function has changed and no compatibility function was provided.
2024-08-17 14:37:35 URL:https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.0-stable/gdextension/extension_api.json [5175722/5175722] -> "/tmp/tmp.qjpaqUWytN" [1]
Godot Engine v4.4.dev.custom_build.1bd740d18 (2024-08-16 22:47:17 UTC) - https://godotengine.org

ERROR: New API lacks base array: enums
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1362)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/draw_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_begin': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
Compatibility to 4.0-stable is broken in the following ways:
Validate extension JSON: API was removed: classes/OpenXRAction
Validate extension JSON: API was removed: classes/OpenXRActionMap
Validate extension JSON: API was removed: classes/OpenXRActionSet
Validate extension JSON: API was removed: classes/OpenXRHand
Validate extension JSON: API was removed: classes/OpenXRInteractionProfile
Validate extension JSON: API was removed: classes/OpenXRInterface
Validate extension JSON: API was removed: classes/OpenXRIPBinding
Validate extension JSON: API was removed: classes/WebXRInterface
2024-08-17 14:37:37 URL:https://raw.githubusercontent.com/godotengine/godot-cpp/godot-4.2-stable/gdextension/extension_api.json [5501700/5501700] -> "/tmp/tmp.5KzMdyqfYe" [1]
Godot Engine v4.4.dev.custom_build.1bd740d18 (2024-08-16 22:47:17 UTC) - https://godotengine.org

ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/draw_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_begin': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
ERROR: Validate extension JSON: Missing field in current API 'classes/RenderingDevice/methods/compute_list_end': arguments. This is a bug.
   at: compare_dict_array (core/extension/extension_api_dump.cpp:1443)
Compatibility to 4.2-stable is broken in the following ways:
Validate extension JSON: API was removed: classes/OpenXRAction
Validate extension JSON: API was removed: classes/OpenXRActionMap
Validate extension JSON: API was removed: classes/OpenXRActionSet
Validate extension JSON: API was removed: classes/OpenXRAPIExtension
Validate extension JSON: API was removed: classes/OpenXRExtensionWrapperExtension
Validate extension JSON: API was removed: classes/OpenXRHand
Validate extension JSON: API was removed: classes/OpenXRInteractionProfile
Validate extension JSON: API was removed: classes/OpenXRInteractionProfileMetadata
Validate extension JSON: API was removed: classes/OpenXRInterface
Validate extension JSON: API was removed: classes/OpenXRIPBinding
Validate extension JSON: API was removed: classes/WebXRInterface
```

There's a lot of "The following validation errors no longer occur" warnings.

IMHO the validation tool seems like is starting to fall short and needs to be improved to handle the complexities of new versions, as it seems to be completely unaware of incremental changes 4.0 -> 4.1 -> 4.2 -> 4.3 -> 4.4.

> The missing hash error is weird, I'll check later.

Yeah :crying_cat_face:

akien-mga commented:
> IMHO the validation tool seems like is starting to fall short and needs to be improved to handle the complexities of new versions, as it seems to be completely unaware of incremental changes 4.0 -> 4.1 -> 4.2 -> 4.3 -> 4.4.

Yeah definitely. I'll see if I can find how to hack things to pass CI so we can merge this, and then how to actually improve the validation tool to be more robust.

akien-mga commented:
So the compat method *is* needed, each change to the API needs to have a compat method, as it's used to generate a hash of the method that GDExtension uses to call it. It's not enough to have source level compatibility in C++, we need binary compatibility.

As for the validation errors caused by the differences between this PR and 4.0-stable and 4.2-stable APIs, I put the message to silence them in their respective `.expected` files, but that's a hack. I'll open an issue so we rework this further, I think we might have bugs either in the validation script, or in the in-engine validation checks.

darksylinc commented:
> So the compat method is needed, each change to the API needs to have a compat method, as it's used to generate a hash of the method that GDExtension uses to call it. **It's not enough to have source level compatibility in C++, we need binary compatibility.**

Somehow I feel this paragraph you just wrote needs to be in https://docs.godotengine.org/en/stable/contributing/development/handling_compatibility_breakages.html

akien-mga commented:
@dsnopek I agree, and tried it, but then our shell script complains that the `4.3.expected` contains lines which are no longer relevant for the 4.4 API. That in itself might be the main bug to fix for this case.

dsnopek commented:
@akien-mga:

> I agree, and tried it, but then our shell script complains that the `4.3.expected` contains lines which are no longer relevant for the 4.4 API. That in itself might be the main bug to fix for this case.

Those messages are technically "warnings" in that it _shouldn't_ cause the CI to fail (at least based on my reading of the shell script). Would some extra non-fatal warnings be acceptable as a workaround for now?

If we decide (as part of the discussion on #95858) that we don't want to have those warnings at all, we can certainly remove them, but I don't personally think that should necessarily block merging this PR.

akien-mga commented:
I did the changes suggested by @dsnopek, and thankfully I also rebased and that let us find that some adjustments are needed after #88199.

```
drivers/metal/rendering_context_driver_metal.mm:73:16: error: allocating an object of abstract class type 'RenderingDeviceDriverMetal' [2]
        return memnew(RenderingDeviceDriverMetal(this));
                       ^
./servers/rendering/rendering_device_driver.h:479:15: note: unimplemented pure virtual method 'shader_destroy_modules' in 'RenderingDeviceDriverMetal' [2]
        virtual void shader_destroy_modules(ShaderID p_shader) = 0;
                      ^
./servers/rendering/rendering_device_driver.h:716:15: note: unimplemented pure virtual method 'command_insert_breadcrumb' in 'RenderingDeviceDriverMetal' [2]
        virtual void command_insert_breadcrumb(CommandBufferID p_cmd_buffer, uint32_t p_data) = 0;
                      ^
```

I suppose those just need to be added to the Metal backend too.

I'll put stubs like for D3D12, and let @stuartcarnie look at it eventually.

akien-mga commented on drivers/metal/rendering_device_driver_metal.mm:
```diff
@@ -2447,6 +2447,10 @@ void deserialize(BufReader &p_reader) {
 	delete obj;
 }
 
+void RenderingDeviceDriverMetal::shader_destroy_modules(ShaderID p_shader) {
+	// TODO.
+}
```
FYI @stuartcarnie - for now this is only called by Vulkan code so it being a no-op shouldn't be an issue (I hope :P).

akien-mga commented on drivers/metal/rendering_device_driver_metal.mm:
```diff
@@ -3541,6 +3545,12 @@ bool isArrayTexture(MTLTextureType p_type) {
 
+#pragma mark - Debug
+
+void RenderingDeviceDriverMetal::command_insert_breadcrumb(CommandBufferID p_cmd_buffer, uint32_t p_data) {
+	// TODO: Implement.
+}
```
FYI @stuartcarnie

akien-mga commented:
Thanks!

darksylinc commented:
WHOA!!! Thanks!

akien-mga commented on drivers/vulkan/rendering_context_driver_vulkan.cpp:
```diff
@@ -40,21 +40,340 @@
+	};
+	VkAllocationCallbacks tracking_callbacks = {
+		// Allocation function
+		nullptr,
+		[](
+				void *p_user_data,
+				size_t size,
+				size_t alignment,
+				VkSystemAllocationScope allocation_scope) -> void * {
```
This lambda seems to not compile properly on arm32.

Here's a log excerpt from a failed Android editor arm32 build:
```
/root/sdk/ndk/23.2.8568313/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ -o editor/editor_string_names.android.editor.arm32.o -c -std=gnu++17 -fno-exceptions -target armv7a-linux-androideabi21 -fpic -ffunction-sections -funwind-tables -fstack-protector-strong -fvisibility=hidden -fno-strict-aliasing -march=armv7-a -mfloat-abi=softfp -ffp-contract=off -O2 -w -DTOOLS_ENABLED -DDEBUG_ENABLED -DNDEBUG -DENGINE_UPDATE_CHECK_ENABLED -DNO_EDITOR_SPLASH -D__ARM_ARCH_7__ -D__ARM_ARCH_7A__ -D__ARM_NEON__ -DANDROID_ENABLED -DUNIX_ENABLED -DVULKAN_ENABLED -DRD_ENABLED -DGLES3_ENABLED -DMINIZIP_ENABLED -DBROTLI_ENABLED -DTHREADS_ENABLED -DCLIPPER2_ENABLED -DZSTD_STATIC_LINKING_ONLY -DUSE_VOLK -DVK_USE_PLATFORM_ANDROID_KHR -DHAVE_MNTENT -Ithirdparty/freetype/include -Ithirdparty/libpng -Ithirdparty/volk -Ithirdparty/vulkan -Ithirdparty/vulkan/include -Ithirdparty/zstd -Ithirdparty/zlib -Ithirdparty/clipper2/include -Ithirdparty/brotli/include -Iplatform/android -I. editor/editor_string_names.cpp
[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:240:3: [0m[0;1;31merror: [0m[1mno viable conversion from '(lambda at drivers/vulkan/rendering_context_driver_vulkan.cpp:240:3)' to 'PFN_vkAllocationFunction' (aka 'void *(*)(void *, unsigned int, unsigned int, VkSystemAllocationScope) __attribute__((pcs("aapcs-vfp")))')[0m
                [](
[0;1;32m                ^~~
[0m[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:240:3: [0m[0;1;30mnote: [0mcandidate function[0m
[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:272:3: [0m[0;1;31merror: [0m[1mno viable conversion from '(lambda at drivers/vulkan/rendering_context_driver_vulkan.cpp:272:3)' to 'PFN_vkReallocationFunction' (aka 'void *(*)(void *, void *, unsigned int, unsigned int, VkSystemAllocationScope) __attribute__((pcs("aapcs-vfp")))')[0m
                [](
[0;1;32m                ^~~
[0m[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:272:3: [0m[0;1;30mnote: [0mcandidate function[0m
[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:306:3: [0m[0;1;31merror: [0m[1mno viable conversion from '(lambda at drivers/vulkan/rendering_context_driver_vulkan.cpp:306:3)' to 'PFN_vkFreeFunction' (aka 'void (*)(void *, void *) __attribute__((pcs("aapcs-vfp")))')[0m
                [](
[0;1;32m                ^~~
[0m[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:306:3: [0m[0;1;30mnote: [0mcandidate function[0m
[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:325:3: [0m[0;1;31merror: [0m[1mno viable conversion from '(lambda at drivers/vulkan/rendering_context_driver_vulkan.cpp:325:3)' to 'PFN_vkInternalAllocationNotification' (aka 'void (*)(void *, unsigned int, VkInternalAllocationType, VkSystemAllocationScope) __attribute__((pcs("aapcs-vfp")))')[0m
                [](
[0;1;32m                ^~~
[0m[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:325:3: [0m[0;1;30mnote: [0mcandidate function[0m
[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:331:3: [0m[0;1;31merror: [0m[1mno viable conversion from '(lambda at drivers/vulkan/rendering_context_driver_vulkan.cpp:331:3)' to 'PFN_vkInternalFreeNotification' (aka 'void (*)(void *, unsigned int, VkInternalAllocationType, VkSystemAllocationScope) __attribute__((pcs("aapcs-vfp")))')[0m
                [](
[0;1;32m                ^~~
[0m[1mdrivers/vulkan/rendering_context_driver_vulkan.cpp:331:3: [0m[0;1;30mnote: [0mcandidate function[0m
/root/sdk/ndk/23.2.8568313/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ -o editor/editor_help.android.editor.arm32.o -c -std=gnu++17 -fno-exceptions -target armv7a-linux-androideabi21 -fpic -ffunction-sections -funwind-tables -fstack-protector-strong -fvisibility=hidden -fno-strict-aliasing -march=armv7-a -mfloat-abi=softfp -ffp-contract=off -O2 -w -DTOOLS_ENABLED -DDEBUG_ENABLED -DNDEBUG -DENGINE_UPDATE_CHECK_ENABLED -DNO_EDITOR_SPLASH -D__ARM_ARCH_7__ -D__ARM_ARCH_7A__ -D__ARM_NEON__ -DANDROID_ENABLED -DUNIX_ENABLED -DVULKAN_ENABLED -DRD_ENABLED -DGLES3_ENABLED -DMINIZIP_ENABLED -DBROTLI_ENABLED -DTHREADS_ENABLED -DCLIPPER2_ENABLED -DZSTD_STATIC_LINKING_ONLY -DUSE_VOLK -DVK_USE_PLATFORM_ANDROID_KHR -DHAVE_MNTENT -Ithirdparty/freetype/include -Ithirdparty/libpng -Ithirdparty/volk -Ithirdparty/vulkan -Ithirdparty/vulkan/include -Ithirdparty/zstd -Ithirdparty/zlib -Ithirdparty/clipper2/include -Ithirdparty/brotli/include -Iplatform/android -I. editor/editor_help.cpp
5 errors generated.
```
(sorry for the ANSI codes mangled in)

darksylinc replied:
That's odd.

The signature for PFN_vkAllocationFunction is:

```cpp
typedef void* (VKAPI_PTR *PFN_vkAllocationFunction)(
    void*                                       pUserData,
    size_t                                      size,
    size_t                                      alignment,
    VkSystemAllocationScope                     allocationScope);
```

Which is defined in godot/thirdparty/vulkan/include/vulkan/vulkan_core.h

And our lambda is:

```cpp
		[](
				void *p_user_data,
				size_t size,
				size_t alignment,
				VkSystemAllocationScope allocation_scope) -> void * {
```

The only thing that comes to mind is that the definition of "size_t" may be different when parsing vulkan_core.h than when parsing the lambda in rendering_context_driver_vulkan.cpp

That's the most suspicious one since Vulkan has a tendency to define everything as uint64, while size_t should be 32-bit in arm32.

Is there an easy way I can repro this? (I have xp with compiling for Android, but I've been compiling for aarch64).

Does the problem go away if you change the size_t in the lambdas for uint64_t ? (and what about uint32_t?)
